library/alloc/src/vec/mod.rs

```diff
@@ -1775,7 +1775,31 @@ impl<T, A: Allocator> Vec<T, A> {
             return;
         }
 
-        /* INVARIANT: vec.len() > read >= write > write-1 >= 0 */
+        // Check if we ever want to remove anything.
+        // This allows to use copy_non_overlapping in next cycle.
+        // And avoids any memory writes if we don't need to remove anything.
+        let mut possible_remove_idx = 1;
+        let start = self.as_mut_ptr();
+        while possible_remove_idx < len {
+            let found_duplicate = unsafe {
+                // SAFETY: possible_remove_idx always in range [1..len)
+                // Note that we start iteration from 1 so we never overflow.
+                let prev = start.add(possible_remove_idx.wrapping_sub(1));
+                let current = start.add(possible_remove_idx);
+                // We explicitly say in docs that references are reversed.
+                same_bucket(&mut *current, &mut *prev)
+            };
+            if found_duplicate {
+                break;
+            }
+            possible_remove_idx += 1;
+        }
+        // Don't need to remove anything.
+        if possible_remove_idx >= len {
+            return;
+        }
```

Comment:
If it can exceed the length then that's problematic for `Vec<()>` since that can have a length of `usize::MAX`.

Reply:
Well, valid indexes are always smaller than length so even in that case there would not be overflow.

Reply:
So the `>` in `>=` isn't really necessary? Ok.